airbyte-cdk/python/airbyte_cdk/sources/file_based/file_types/csv_parser.py

```diff
@@ -47,78 +52,137 @@ async def infer_schema(
+        """
+        reader = csv.reader(fp, dialect=dialect_name)  # type: ignore
+        number_of_columns = len(next(reader))  # type: ignore
+        return [f"f{i}" for i in range(number_of_columns)]
+
+    @staticmethod
+    def _skip_rows(fp: IOBase, rows_to_skip: int) -> None:
+        """
+        Skip rows before the header. This has to be done on the file object itself, not the reader
+        """
+        for _ in range(rows_to_skip):
+            fp.readline()
+
+
+class CsvParser(FileTypeParser):
+    _MAX_BYTES_PER_FILE_FOR_SCHEMA_INFERENCE = 1_000_000
+
+    def __init__(self, csv_reader: Optional[_CsvReader] = None):
+        self._csv_reader = csv_reader if csv_reader else _CsvReader()
+
+    async def infer_schema(
+        self,
+        config: FileBasedStreamConfig,
+        file: RemoteFile,
+        stream_reader: AbstractFileBasedStreamReader,
+        logger: logging.Logger,
+    ) -> SchemaType:
+        input_schema = config.get_input_schema()
+        if input_schema:
+            return input_schema
+
+        # todo: the existing InMemoryFilesSource.open_file() test source doesn't currently require an encoding, but actual
+        #  sources will likely require one. Rather than modify the interface now we can wait until the real use case
+        config_format = _extract_format(config)
+        type_inferrer_by_field: Dict[str, _TypeInferrer] = defaultdict(
+            lambda: _JsonTypeInferrer(
+                config_format.true_values,
+                config_format.false_values,
+                config_format.null_values,
+                config_format.inference_type == InferenceType.PRIMITIVE_AND_COMPLEX_TYPES,
+            )
+            if config_format.inference_type != InferenceType.NONE
+            else _DisabledTypeInferrer()
+        )
+        data_generator = self._csv_reader.read_data(config, file, stream_reader, logger, self.file_read_mode)
+        read_bytes = 0
+        for row in data_generator:
+            for header, value in row.items():
+                type_inferrer_by_field[header].add_value(value)
+                # This is not accurate as a representation of how many bytes were read because csv does some processing on the actual value
+                # before returning. Given we would like to be more accurate, we could wrap the IO file using a decorator
+                read_bytes += len(value)
+            read_bytes += len(row) - 1  # for separators
+            if read_bytes >= self._MAX_BYTES_PER_FILE_FOR_SCHEMA_INFERENCE:
+                break
+
+        schema = {header.strip(): {"type": type_inferred.infer()} for header, type_inferred in type_inferrer_by_field.items()}
+        data_generator.close()
+        return schema
 
     def parse_records(
         self,
         config: FileBasedStreamConfig,
         file: RemoteFile,
         stream_reader: AbstractFileBasedStreamReader,
         logger: logging.Logger,
+        discovered_schema: Optional[SchemaType],
```

Comment:
I'm not sure discovered_schema is an `Optional[SchemaType]`

SchemaType resolves to `Mapping[str, Mapping[Union[List[str], Literal["string"], str]]`

But on [line184](https://github.com/airbytehq/airbyte/pull/29099/files#diff-2ac011610fd803eed6b2f7f1369e9c9ecc6d99b01ebfb475781eec212d081064R184), we access it as if it was a 

`Mapping[str, Mapping[str, Any]]`

I think updating JsonSchemaSupportedType is enough, but it's possible we're reaching the limits of what we can express without defining the structure of the object we expect (eg. it should have a properties field)

Reply:
Typing primitive types is complex...